In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

In [6]:
# Example RF dataset
import pandas as pd

# Open the CSV file instead of creating new data
data = pd.read_csv('radio_propagation_data.csv')

# data.head()
data.shape

(500, 5)

In [7]:
X = data[['distance_km', 'frequency_mhz', 'tx_power_dbm', 'antenna_height_m']]
y = data['rsrp_dbm']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

In [8]:
model = RandomForestRegressor(
    n_estimators=100,
    random_state=42
)

model.fit(X_train, y_train)

predictions = model.predict(X_test)

In [9]:
mae = mean_absolute_error(y_test, predictions)
mse = mean_squared_error(y_test, predictions)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, predictions)

print("Predicted RSRP:", predictions)
print("Actual RSRP:", list(y_test))
print("MAE:", mae)
print("MSE:", mse)
print("RMSE:", rmse)
print("R^2 Score:", r2)

Predicted RSRP: [-61.25       -41.73       -63.6        -47.35       -43.58
 -62.01       -63.82       -37.88       -34.68       -60.33
 -20.05       -49.18       -63.04       -41.7        -63.14
 -63.85       -73.5        -26.58       -57.72333333 -61.82
 -63.11       -74.4        -63.45       -68.28       -33.37
 -62.5        -68.68       -53.04       -35.36       -62.19
 -73.06       -58.26       -70.01       -50.67       -19.48
 -63.14       -67.79       -41.3        -56.12       -31.91
 -64.06       -31.6        -72.94       -53.79       -25.69
 -41.62       -19.2652381  -23.435      -69.13       -22.39
 -34.49       -49.97       -46.22       -61.22       -56.78
 -43.92       -47.26       -19.35757143 -52.96       -44.62
 -67.02       -57.49       -62.79       -61.9        -34.41
 -35.43       -63.39       -26.42       -51.83       -62.49
 -61.02       -62.21       -61.90611111 -57.96111111 -59.79
 -63.58       -56.274      -35.06       -61.77       -61.45
 -54.9        -36.06    

# Regression Model Performance Metrics

## Overview
These four metrics evaluate the performance of a regression model predicting RSRP (Reference Signal Received Power) values in a cellular network propagation model.

---

## 1. Mean Absolute Error (MAE): 2.79 dBm

### Formula
$$MAE = \frac{1}{n}\sum_{i=1}^{n}|y_i - \hat{y}_i|$$

### Explanation
- **What it measures**: The average absolute difference between predicted and actual values
- **Interpretation**: On average, predictions are off by **2.79 dBm**
- **Advantages**: 
  - Easy to interpret (same units as target variable)
  - Robust to outliers (doesn't square errors)
- **Disadvantages**: 
  - Not differentiable (can't be used in some optimization algorithms)
  - Less sensitive to large errors

---

## 2. Mean Squared Error (MSE): 12.28 dBm²

### Formula
$$MSE = \frac{1}{n}\sum_{i=1}^{n}(y_i - \hat{y}_i)^2$$

### Explanation
- **What it measures**: The average of squared differences between predicted and actual values
- **Interpretation**: Average squared error is **12.28 dBm²**
- **Advantages**: 
  - Differentiable (useful for gradient-based optimization)
  - Heavily penalizes large errors (encourages fewer big mistakes)
- **Disadvantages**: 
  - Units are squared (not directly interpretable)
  - Sensitive to outliers (large errors are magnified)

---

## 3. Root Mean Squared Error (RMSE): 3.50 dBm

### Formula
$$RMSE = \sqrt{MSE} = \sqrt{\frac{1}{n}\sum_{i=1}^{n}(y_i - \hat{y}_i)^2}$$

### Explanation
- **What it measures**: The typical error magnitude in original units
- **Interpretation**: Typical prediction error is approximately **3.50 dBm**
- **Comparison**: RMSE > MAE (3.50 > 2.79) indicates:
  - Some larger errors exist
  - Difference of 0.71 dBm suggests error distribution is relatively normal
- **Advantages**: 
  - Same units as target variable (easily interpretable)
  - Good for comparing models
- **Disadvantages**: 
  - Still sensitive to outliers (due to squared errors)

### Error Distribution Rules (Normal Distribution)
- **68-95-99.7 Rule** (if errors are normally distributed):
  - 68% of predictions: within ±3.50 dBm
  - 95% of predictions: within ±7.00 dBm
  - 99.7% of predictions: within ±10.50 dBm

---

## 4. R² Score (Coefficient of Determination): 0.9462

### Formula
$$R^2 = 1 - \frac{SS_{residual}}{SS_{total}} = 1 - \frac{\sum_{i=1}^{n}(y_i - \hat{y}_i)^2}{\sum_{i=1}^{n}(y_i - \bar{y})^2}$$

Where:
- $SS_{residual}$ = Sum of squared residuals (unexplained variance)
- $SS_{total}$ = Total sum of squares (total variance)
- $\bar{y}$ = Mean of actual values

### Explanation
- **What it measures**: Proportion of variance in the dependent variable explained by the model
- **Interpretation**: Model explains **94.62%** of the variance in RSRP values
- **Only 5.38%** of variance remains unexplained
- **Range**: 0 to 1 (0% to 100%)
  - 1.0 = Perfect predictions
  - 0.0 = Model predicts the mean only
  - Negative values = Model performs worse than the mean

### R² Interpretation Guidelines

| R² Value | Performance | Explanation |
|----------|-------------|-------------|
| 0.90 - 1.00 | **Excellent** | Model captures almost all variation |
| 0.70 - 0.89 | **Good** | Strong predictive power |
| 0.50 - 0.69 | **Moderate** | Adequate, room for improvement |
| 0.30 - 0.49 | **Weak** | Limited explanatory power |
| < 0.30 | **Poor** | Model fails to explain variance |

---

## Comprehensive Performance Summary

### Overall Assessment: **EXCELLENT** ✅

| Metric | Value | Unit | Best Value | Performance |
|--------|-------|------|------------|-------------|
| **MAE** | 2.79 | dBm | 0 (lower is better) | ✅ Excellent |
| **MSE** | 12.28 | dBm² | 0 (lower is better) | ✅ Excellent |
| **RMSE** | 3.50 | dBm | 0 (lower is better) | ✅ Excellent |
| **R²** | 0.9462 | - | 1.0 (higher is better) | ✅ Excellent |

---
